Fluid typography: extract logic to derive fluid typography settings

ramonjd opened the pull request:
Tracking issue: 
- https://github.com/WordPress/gutenberg/issues/44888

## What?
This commit updates the utilities to house logic to get the typography settings. See context: https://github.com/WordPress/gutenberg/pull/51146#discussion_r1212533596

Adds tests as well.

## Why?

Because there's a little tap dance that's needed to check whether fluid typography is enabled via boolean or object props, and also whether a maximum viewport size value can be extracted from layout settings (if fluid typography is enabled).

## How?
A few `if`s and `return`s

## Testing Instructions
The new and existing tests should pass.

```shell
npm run test:unit packages/block-editor/src/components/global-styles/test/typography-utils.js
npm run test:unit packages/block-editor/src/components/global-styles/test/use-global-styles-output.js
```

ramonjd committed:
This commit updates the utilities to house logic to get the typography settings.
Why is this required? Because there's a little tap dance that's needed to check whether fluid typography is enabled via boolean or object, and also whether a maximum viewport size value can be extracted from layout settings.
Adds tests as well.

andrewserong commented on packages/block-editor/src/components/global-styles/typography-utils.js:
```diff
@@ -85,3 +76,36 @@ export function getTypographyFontSizeValue( preset, typographySettings ) {
+	const fluidSettings = typographySettings?.fluid;
+	return (
+		true === fluidSettings ||
+		( typeof fluidSettings === 'object' &&
+			Object.keys( fluidSettings ).length > 0 )
```
Not a blocker since this is just a copy + paste from the existing code. Also, it's probably an edge case because I doubt anyone would do it, but if `typographySettings?.fluid` is set to `null` then it'll be passed to `Object.keys()` and potentially throw an error (since `typeof` `null` is `object`). I ran into another error unrelated to this PR over in https://github.com/WordPress/gutenberg/pull/51252 so it's on my mind 😄

Would it work to update this line to `fluidSettings && fluidSettings === 'object' && Object.keys( fluidSettings ).length > 0`?

ramonjd replied:
Much better. I like!

andrewserong commented on packages/block-editor/src/components/global-styles/typography-utils.js:
```diff
@@ -85,3 +76,36 @@ export function getTypographyFontSizeValue( preset, typographySettings ) {
+ * @param {Object} settings            Theme.json settings
+ * @param {Object} settings.typography Theme.json typography settings
+ * @param {Object} settings.layout     Theme.json layout settings
+ * @return {{fluid: (*&{maxViewPortWidth})}|{fluid: *}} Fluid typography settings
```
Nit: it looks like the right hand type is more permissive than the left one, so I'm wondering if the `*&{maxViewPortWidth})}` is necessary? In VSCode if I assign `let item = getFluidTypographyOptionsFromSettings( tree?settings );` the result appears to be `any` for `fluid` in all cases:

<img width="717" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/672b9105-dcd7-495a-86e0-2e1c82a72b5c">

ramonjd replied:
Ugh, this was my IDE. Please believe me! 😄 Thanks for spotting it 🙇 

I'd already written a type object above `TypographySettings`. Auto format killed it!

andrewserong reviewed (approved):
This is testing nicely for me @ramonjd, thanks for following up!

Just left a couple of non-blocking nits, but this LGTM 🙂

noisysocks commented on packages/block-editor/src/components/global-styles/typography-utils.js:
```diff
@@ -45,24 +45,15 @@ import { getComputedFluidTypographyValue } from '../font-sizes/fluid-utils';
 export function getTypographyFontSizeValue( preset, typographySettings ) {
```
Dumb nit but can we rename `typographySettings` here to `typographyOptions` to indicate that you're supposed to pass it the return value from `getFluidTypographyOptionsFromSettings`?

`settings` → `typographySettings` → `typographyOptions`

ramonjd replied:
Good call. Will do

ramonjd committed:
- rename typographySettings here to typographyOptions to indicate that we're supposed to pass it the return value from getFluidTypographyOptionsFromSettings
- use TypographySettings as object return type
- test for null values in isFluidTypographyEnabled check
